hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/MapIndexScanP.java

```diff
@@ -0,0 +1,431 @@
+                    extremeIndex = i;
+                    extreme = s.currentRow;
+                }
+            }
+
+            if (extremeIndex < 0) {
+                assert splits.isEmpty();
+                return true;
+            }
+
+            pendingItem = extreme;
+            if (lastSentItem != null) {
+                assert ((Comparable) pendingItem[1]).compareTo(lastSentItem[1]) ==
+                        metadata.getComparator().compare(pendingItem, lastSentItem) :
+                        "seen : " + pendingItem[3] + ", last sent : " + lastSentItem[3] + ", " +
+                                "bytes value : " + pendingItem[16] + ", last sent : " + lastSentItem[16];
+            }
+            splits.get(extremeIndex).remove();
+        }
+    }
+
+    /**
+     * Perform splitting of a {@link Split} after receiving {@link MissingPartitionException}.
+     * Method gets current partition table and proceeds with following procedure:
+     * <p>
+     * It splits the partitions assigned to the split according to the new
+     * partition owner into disjoint sets, one for each owner.
+     *
+     * @param split the split to split
+     * @return collection of new split units
+     */
+    private List<Split> splitOnMigration(Split split) {
+        IndexIterationPointer[] lastPointers = split.pointers;
+        InternalPartitionService partitionService = getNodeEngine(hazelcastInstance).getPartitionService();
+        PrimitiveIterator.OfInt partitionIterator = split.partitions.intIterator();
+        Map<Address, Split> newSplits = new HashMap<>();
+        while (partitionIterator.hasNext()) {
+            int partitionId = partitionIterator.nextInt();
+            Address owner = partitionService.getPartition(partitionId).getOwnerOrNull();
+            // TODO: does it possible that owner is NULL?
+            newSplits.computeIfAbsent(owner, x -> new Split(
+                    new PartitionIdSet(partitionService.getPartitionCount()), owner, lastPointers)
+            ).partitions.add(partitionId);
+        }
+        return new ArrayList<>(newSplits.values());
+    }
+
+    private IndexIterationPointer[] filtersToPointers(@Nonnull IndexFilter filter, boolean descending) {
+        return IndexIterationPointer.createFromIndexFilter(filter, descending, evalContext);
+    }
+
+    private Object[] doFullProjectionAndFilter(@Nonnull QueryableEntry<?, ?> entry) {
+        row.setKeyValue(entry.getKey(), entry.getKeyData(), entry.getValue(), entry.getValueData());
+        if (metadata.getRemainingFilter() != null
+                && TernaryLogic.isNotTrue(metadata.getRemainingFilter().evalTop(row, evalContext))) {
+            return null;
+        }
+
+        Object[] row = new Object[metadata.getFullProjection().size()];
+        for (int j = 0; j < metadata.getFullProjection().size(); j++) {
+            row[j] = evaluate(metadata.getFullProjection().get(j), this.row, evalContext);
+        }
+        return row;
```

Comment:
Instead of creating another projecting vertex in `IMapSqlConnector`, could not we apply the required projection here?